File: blueteeth-sample/src/main/java/com/robotpajamas/android/blueteeth/DeviceActivity.java
package com.robotpajamas.android.blueteeth;

import android.app.Activity;
import android.os.Bundle;
import android.widget.Button;
import android.widget.ScrollView;
import android.widget.TextView;

import com.robotpajamas.android.blueteeth.peripherals.SamplePeripheral;
import com.robotpajamas.blueteeth.BlueteethManager;
import com.robotpajamas.blueteeth.BlueteethResponse;

import java.util.Arrays;

import butterknife.Bind;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class DeviceActivity extends Activity {

    private SamplePeripheral mSamplePeripheral;
    private boolean mIsConnected;

    @Bind(R.id.scrollview)
    ScrollView mScrollView;

    @Bind(R.id.textview_console)
    TextView mConsoleTextView;

    @Bind(R.id.button_connect)
    Button mConnectionButton;

    @Bind(R.id.button_read_counter)
    Button mReadCounterButton;

    @Bind(R.id.button_write)
    Button mWriteButton;

    @Bind(R.id.button_write_no_response)
    Button mWriteNoResponseButton;

    @OnClick(R.id.button_clear)
    void clearConsole() {
        mConsoleTextView.setText("");
    }

    @OnClick(R.id.button_connect)
    void connect() {
        if (mIsConnected) {
            updateReceivedData(String.format("Attempting to disconnect from %s - %s...", mSamplePeripheral.getName(), mSamplePeripheral.getMacAddress()));
            mSamplePeripheral.disconnect(isConnected -> {
                updateReceivedData("Connection Status: " + Boolean.toString(isConnected) + "\n");
                mIsConnected = isConnected;
                runOnUiThread(mConnectionRunnable);
            });
        } else {
            updateReceivedData(String.format("Attempting to connect to  %s - %s...", mSamplePeripheral.getName(), mSamplePeripheral.getMacAddress()));
            mSamplePeripheral.connect(true, isConnected -> {
                updateReceivedData("Connection Status: " + Boolean.toString(isConnected));
                mIsConnected = isConnected;
                runOnUiThread(mConnectionRunnable);
            });
        }
    }

    Runnable mConnectionRunnable = new Runnable() {
        @Override
        public void run() {
            if (mIsConnected) {
                mConnectionButton.setText(R.string.disconnect);
                mReadCounterButton.setEnabled(true);
                mWriteButton.setEnabled(true);
                mWriteNoResponseButton.setEnabled(true);
            } else {
                mConnectionButton.setText(R.string.connect);
                mReadCounterButton.setEnabled(false);
                mWriteButton.setEnabled(false);
                mWriteNoResponseButton.setEnabled(false);
            }
        }
    };

    @OnClick(R.id.button_read_counter)
    void readCounter() {
        updateReceivedData("Attempting to Read Counter ...");
        mSamplePeripheral.readCounter((response, data) -> {
            if (response != BlueteethResponse.NO_ERROR) {
                updateReceivedData("Read error... " + response.name());
                return;
            }

            updateReceivedData(Arrays.toString(data));
        });
    }

    @OnClick(R.id.button_write)
    void write() {
        updateReceivedData("Attempting to Reset Counter ...");
        mSamplePeripheral.writeCounter((byte) 42, response -> {
            if (response != BlueteethResponse.NO_ERROR) {
                updateReceivedData("Write error... " + response.name());
                return;
            }
            updateReceivedData("Counter characteristic reset to 42");
        });
    }

    @OnClick(R.id.button_write_no_response)
    void writeNoResponse() {
        updateReceivedData("Resetting Counter with No Response ...");
        mSamplePeripheral.writeNoResponseCounter((byte) 42);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_device);
        ButterKnife.bind(this);

        String macAddress = getIntent().getStringExtra(getString(R.string.extra_mac_address));
        mSamplePeripheral = new SamplePeripheral(BlueteethManager.with(this).getPeripheral(macAddress));
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mSamplePeripheral.close();
    }

    private void updateReceivedData(String message) {
        runOnUiThread(() -> {
            mConsoleTextView.append(message + "\n");
            mScrollView.smoothScrollTo(0, mConsoleTextView.getBottom());
        });
    }
}

File: blueteeth-sample/src/main/kotlin/com/robotpajamas/android/blueteeth/DeviceActivity.kt
package com.robotpajamas.android.blueteeth

//import com.robotpajamas.android.blueteeth.peripherals.SamplePeripheral;

import android.app.Activity
import android.databinding.DataBindingUtil
import android.os.Bundle
import android.widget.Button
import android.widget.ScrollView
import android.widget.TextView
import com.robotpajamas.android.blueteeth.databinding.ActivityDeviceBinding
import com.robotpajamas.blueteeth.BlueteethDevice

class DeviceActivity : Activity() {

    lateinit var binding: ActivityDeviceBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_device)

        val macAddress = intent.getStringExtra(getString(R.string.extra_mac_address))
        //        mSamplePeripheral = new SamplePeripheral(Blueteeth.INSTANCE.getPeripheral(macAddress));
    }

    override fun onDestroy() {
        super.onDestroy()
        mSamplePeripheral?.close()
    }

    private val mSamplePeripheral: BlueteethDevice? = null
    private val mIsConnected: Boolean = false

    //    @BindView(R.id.scrollview)
    internal var mScrollView: ScrollView? = null

    //    @BindView(R.id.textview_console)
    internal var mConsoleTextView: TextView? = null

    //    @BindView(R.id.button_connect)
    internal var mConnectionButton: Button? = null

    //    @BindView(R.id.button_write)
    internal var mWriteButton: Button? = null

    //    @BindView(R.id.button_write_no_response)
    internal var mWriteNoResponseButton: Button? = null

    //    @BindView(R.id.button_read_counter)
    internal var mReadCounterButton: Button? = null

    //    @BindView(R.id.button_toggle_notify)
    internal var mToggleNotifyButton: Button? = null

    //    @BindView(R.id.button_toggle_indicate)
    internal var mToggleIndicateButton: Button? = null

    internal var mConnectionRunnable: Runnable = Runnable {
        if (mIsConnected) {
            mConnectionButton!!.setText(R.string.disconnect)
            mWriteButton!!.isEnabled = true
            mWriteNoResponseButton!!.isEnabled = true
            mReadCounterButton!!.isEnabled = true
            mToggleNotifyButton!!.isEnabled = true
            mToggleIndicateButton!!.isEnabled = true
        } else {
            mConnectionButton!!.setText(R.string.connect)
            mWriteButton!!.isEnabled = false
            mWriteNoResponseButton!!.isEnabled = false
            mReadCounterButton!!.isEnabled = false
            mToggleNotifyButton!!.isEnabled = false
            mToggleIndicateButton!!.isEnabled = false
        }
    }

    private var mNotifyEnabled = false

    //    @OnClick(R.id.button_clear)
    internal fun clearConsole() {
        mConsoleTextView!!.text = ""
    }

    //    @OnClick(R.id.button_connect)
    internal fun connect() {
        if (mIsConnected) {
            updateReceivedData(String.format("Attempting to disconnect from %s - %s...", mSamplePeripheral!!.name, mSamplePeripheral.macAddress))
            //            mSamplePeripheral.disconnect(isConnected -> {
            //                updateReceivedData("Connection Status: " + Boolean.toString(isConnected) + "\n");
            //                mIsConnected = isConnected;
            //                runOnUiThread(mConnectionRunnable);
            //            });
        } else {
            updateReceivedData(String.format("Attempting to connect to  %s - %s...", mSamplePeripheral!!.name, mSamplePeripheral.macAddress))
            //            mSamplePeripheral.connect(false, isConnected -> {
            //                updateReceivedData("Connection Status: " + Boolean.toString(isConnected));
            //                mIsConnected = isConnected;
            //                runOnUiThread(mConnectionRunnable);
            //            });
        }
    }

    //    @OnClick(R.id.button_read_counter)
    internal fun readCounter() {
        updateReceivedData("Attempting to Read Counter ...")
        //        mSamplePeripheral.readCounter((response, data) -> {
        //            if (response != BlueteethResponse.NO_ERROR) {
        //                updateReceivedData("Read error... " + response.name());
        //                return;
        //            }
        //            updateReceivedData(Arrays.toString(data));
        //        });
    }

    //    @OnClick(R.id.button_toggle_notify)
    internal fun toggleNotify() {
        mNotifyEnabled = !mNotifyEnabled
        updateReceivedData("Toggle notifications ...")
        //        mSamplePeripheral.toggleNotification(mNotifyEnabled, (response, data) -> {
        //            if (response != BlueteethResponse.NO_ERROR) {
        //                updateReceivedData("Notification error... " + response.name());
        //                return;
        //            }
        //            updateReceivedData(Arrays.toString(data));
        //        });
    }

    //    @OnClick(R.id.button_write)
    internal fun write() {
        updateReceivedData("Attempting to Reset Counter ...")
        //        mSamplePeripheral.writeCounter((byte) 42, response -> {
        //            if (response != BlueteethResponse.NO_ERROR) {
        //                updateReceivedData("Write error... " + response.name());
        //                return;
        //            }
        //            updateReceivedData("Counter characteristic reset to 42");
        //        });
    }

    //    @OnClick(R.id.button_write_no_response)
    internal fun writeNoResponse() {
        updateReceivedData("Resetting Counter with No Response ...")
        //        mSamplePeripheral.writeNoResponseCounter((byte) 42);
    }


    private fun updateReceivedData(message: String) {
        //        runOnUiThread(() -> {
        //            mConsoleTextView.append(message + "\n");
        //            mScrollView.smoothScrollTo(0, mConsoleTextView.getBottom());
        //        });
    }
}
